Implement CreateTasks using slice-based batching

sevein opened the pull request:
Adopt slice-based CreateTasks across ingest workflows, superseding PR #1438.

sevein committed:
Implement CreateTasks

sevein committed:
Use CreateTasks in CreateAIPActivity (a3m)

sevein committed:
Use CreateTasks in JobTracker (am)

sevein committed:
Use CreateTasks in SavePreprocessingTasksActivity

sevein committed:
Use numeric UID/GID in Dockerfile for cross-platform builds

sevein committed:
Enable copyloopvar linter

jraddaoui commented on internal/a3m/a3m.go:
```diff
@@ -175,12 +186,7 @@ func saveTasks(
-		transferservice.Job_STATUS_PROCESSING:  enums.TaskStatusInProgress,
-		transferservice.Job_STATUS_FAILED:      enums.TaskStatusError,
-	}
+	tasks := make([]*datatypes.Task, 0, len(jobs))
```
Same as in the other PR, nitpicking, use `make([]*datatypes.Task, len(jobs))` and add using the index from the loop below.

jraddaoui commented on internal/a3m/a3m.go:
```diff
@@ -165,6 +168,14 @@ func (a *CreateAIPActivity) Execute(
 }
 
+// taskStatusesByA3mJobStatuses maps a3m job statuses to Enduro task statuses.
+var taskStatusesByA3mJobStatuses = map[transferservice.Job_Status]enums.TaskStatus{
```
I think the name we better before, or what I'd understand better, `a3mJobToTaskStatuses`. Not a big deal for sure.

jraddaoui commented on internal/ingest/task.go:
```diff
@@ -25,6 +25,26 @@ func (svc *ingestImpl) CreateTask(ctx context.Context, task *datatypes.Task) err
+
+	err := svc.perSvc.CreateTasks(ctx, tasks)
+	if err != nil {
+		return fmt.Errorf("tasks: create: %v", err)
```
Feels wrong to change the error prefix to a plural version, maybe `task: create bulk` and change the one above to `task: create one` or leave as it is. Or just remove the prefix from the error messages and use the function name directly. Not a big deal either.

sevein replied:
Makes sense. I went with `tasks: bulk create: %v` so it sounds like we're performing a creation op in bulk as opposed to a "bulk" object.

jraddaoui commented on internal/persistence/ent/client/task.go:
```diff
@@ -13,59 +14,116 @@ import (
-	if task.WorkflowUUID == uuid.Nil {
-		return newRequiredFieldError("WorkflowUUID")
+	if task == nil {
+		return newRequiredFieldError("task")
```
You are looking for this in the loop from CreateTasks, I think it would be okay to remove it here.

jraddaoui commented on internal/persistence/telemetry_test.go:
```diff
@@ -17,6 +21,101 @@ import (
 	"github.com/artefactual-sdps/enduro/internal/persistence/fake"
 )
 
+func newTracerWithRecorder(t *testing.T) (*tracetest.SpanRecorder, *otelsdktrace.TracerProvider) {
+	t.Helper()
+
+	recorder := tracetest.NewSpanRecorder()
+	tp := otelsdktrace.NewTracerProvider(otelsdktrace.WithSpanProcessor(recorder))
+	t.Cleanup(func() { _ = tp.Shutdown(context.Background()) })
+
+	return recorder, tp
+}
```
Nice!

jraddaoui commented on internal/persistence/ent/client/task_test.go:
```diff
@@ -128,6 +133,146 @@ func TestCreateTask(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	dsn := fmt.Sprintf("file:%s?mode=memory&cache=shared&_fk=1", t.Name())
+
+	sqlDB, err := sql.Open("sqlite3", dsn)
+	assert.NilError(t, err)
+	t.Cleanup(func() { sqlDB.Close() })
+
+	drv := entsql.OpenDB(dialect.SQLite, sqlDB)
+	cdrv := &countingDriver{Driver: drv}
+
+	entc := enttest.NewClient(t, enttest.WithOptions(db.Driver(cdrv)))
+	t.Cleanup(func() { entc.Close() })
+
+	svc := entclient.New(logr.Discard(), entc)
+
+	sip, err := createSIP(t, entc, "Test SIP", enums.SIPStatusIngested)
+	assert.NilError(t, err)
+	_, err = createWorkflow(t, entc, sip.ID, enums.WorkflowStatusDone)
+	assert.NilError(t, err)
+
+	// Ignore INSERTs performed while seeding fixtures.
+	cdrv.reset()
+
+	tasks := make([]*datatypes.Task, 400)
+	for i := range tasks {
+		tasks[i] = &datatypes.Task{
+			UUID:         uuid.New(),
+			Name:         fmt.Sprintf("Task %03d", i),
+			Status:       enums.TaskStatusInProgress,
+			WorkflowUUID: wUUID,
+		}
+	}
+
+	assert.NilError(t, svc.CreateTasks(ctx, tasks))
+	assert.Equal(t, 1, cdrv.begins)
+	assert.Equal(t, 2, len(cdrv.execs))
+}
+
```
I think the test driver approach adds too much complexity for what it does. I think we could live with a normal test, checking that `CreateTasks` is called twice. If you still want to maintain a version of this test, I think it would be simpler with a logging hook, ChatGPT-Codex provided this solution:

```go
// ...existing code...
func TestCreateTasksUsesBatching(t *testing.T) {
    t.Parallel()

    ctx := context.Background()
    dsn := fmt.Sprintf("file:%s?mode=memory&cache=shared&_fk=1", t.Name())

    sqlDB, err := sql.Open("sqlite3", dsn)
    assert.NilError(t, err)
    t.Cleanup(func() { sqlDB.Close() })

    var begins, inserts int
    entc := enttest.NewClient(
        t,
        enttest.WithOptions(
            db.Driver(entsql.OpenDB(dialect.SQLite, sqlDB)),
            db.Log(func(args ...any) { // hook: every Exec/Query goes through here
                q := strings.ToUpper(fmt.Sprint(args...))
                if strings.Contains(q, "BEGIN") {
                    begins++
                }
                if strings.Contains(q, "INSERT") {
                    inserts++
                }
            }),
        ),
    )
    t.Cleanup(func() { entc.Close() })

    svc := entclient.New(logr.Discard(), entc)
    // ...existing code to seed SIP/workflow and build tasks...
    assert.NilError(t, svc.CreateTasks(ctx, tasks))
    assert.Equal(t, 1, begins)
    assert.Equal(t, 2, inserts)
}
// ...existing code...
```

sevein replied:
Good call.

jraddaoui reviewed (changes requested):
Thanks @sevein!

sevein committed:
Update: address feedback

sevein committed:
Make poller test wait reliably for create event

It should avoid this from happening again:

  === Failed
  === FAIL: internal/filenotify TestPollerEvent (5.02s)
      filenotify_test.go:49: timeout waiting for event CREATE

jraddaoui commented on internal/filenotify/filenotify_test.go:
```diff
@@ -45,22 +45,26 @@ func TestPollerEvent(t *testing.T) {
+			if e.Op == eType {
+				return poll.Success()
+			}
+			return poll.Continue("got wrong event type, expected %q: %v", eType, e.Op)
+		case err := <-w.Errors():
+			return poll.Error(err)
+		default:
+			return poll.Continue("no event yet")
 		}
-	case e := <-w.Errors():
-		err = fmt.Errorf("got unexpected error waiting for events %v: %v", eType, e)
-	case <-time.After(pollInterval * 2):
-		err = fmt.Errorf("timeout waiting for event %v", eType)
-	}
-	return err
+	},
+		poll.WithTimeout(pollInterval*20),
+		poll.WithDelay(pollInterval),
+	)
```
Nice!

jraddaoui reviewed (approved):
Thanks @sevein! I think I've never seen the poller test error, but the changes on it LGTM.
